File: packages/implementations/OneLang-Reflect-v0.1/native/OneReflect.py
classes = {}

def get_class(obj):
    return classes.get(name_key(obj.__class__.__name__))

def get_class_by_name(name):
    return classes.get(name_key(name))

def setup_class(cls):
    classes[name_key(cls.name)] = cls

def name_key(name):
    return name.lower().replace("_", "")

class Class:
    def __init__(self, typeObj, fields, methods):
        self.typeObj = typeObj
        self.name = typeObj.__name__
        
        self.fields = {}
        for field in fields:
            field.cls = self
            self.fields[name_key(field.name)] = field
        
        self.methods = {}
        for method in methods:
            method.cls = self
            self.methods[name_key(method.name)] = method
    
    def get_field(self, name):
        return self.fields.get(name_key(name))
        
    def get_method(self, name):
        return self.methods.get(name_key(name))

    def get_fields(self):
        return self.fields.values()

    def get_methods(self):
        return self.methods.values()

class Field:
    def __init__(self, name, is_static, type):
        self.cls = None
        self.name = name
        self.is_static = is_static
        self.type = type
    
    def get_value(self, obj):
        realObj = self.cls.typeObj if self.is_static else obj
        return getattr(realObj, self.name)
    
    def set_value(self, obj, value):
        realObj = self.cls.typeObj if self.is_static else obj
        setattr(realObj, self.name, value)

class Method:
    def __init__(self, name, is_static, return_type, args):
        self.cls = None
        self.name = name
        self.is_static = is_static
        self.return_type = return_type
        self.args = args
    
    def call(self, obj, args):
        realObj = self.cls.typeObj if self.is_static else obj
        method = getattr(realObj, self.name)
        return method(*args)

class MethodArgument:
    def __init__(self, name, type):
        self.name = name
        self.type = type

File: packages/implementations/OneLang-Reflect-v0.1/native/OneReflect.rb
class OneReflect
    @classes = {}

    class << self
        attr_accessor :classes
    end        

    def self.get_class(obj)
        return get_class_by_name(obj.class.name)
    end

    def self.get_class_by_name(name)
        return self::classes[OneReflect.name_key(name)]
    end

    def self.setup_class(cls)
        self::classes[OneReflect.name_key(cls.name)] = cls
    end

    def self.name_key(name)
        return name.downcase.gsub(/_/, "")
    end

    class Class
        attr_accessor(:type_obj, :name, :fields, :methods)

        def initialize(type_obj, fields, methods)
            @type_obj = type_obj
            @name = type_obj.name
            
            @fields = {}
            for field in fields
                field.cls = self
                @fields[OneReflect.name_key(field.name)] = field
            end

            @methods = {}
            for method in methods
                method.cls = self
                @methods[OneReflect.name_key(method.name)] = method
            end
        end

        def get_field(name)
            return self.fields[OneReflect.name_key(name)]
        end

        def get_method(name)
            return self.methods[OneReflect.name_key(name)]
        end
    
        def get_fields()
            return self.fields.values
        end

        def get_methods()
            return self.methods.values
        end
            
    end

    class Field
        attr_accessor(:cls, :name, :is_static, :type)

        def initialize(name, is_static, type)
            @name = name
            @is_static = is_static
            @type = type
        end

        def get_value(obj)
            realObj = @is_static ? @cls.type_obj : obj;
            return realObj.instance_variable_get("@" + @name)
        end

        def set_value(obj, value)
            realObj = @is_static ? @cls.type_obj : obj;
            realObj.instance_variable_set("@" + @name, value)
        end
    end

    class Method
        attr_accessor(:cls, :name, :is_static, :return_type, :args)

        def initialize(name, is_static, return_type, args)
            @name = name
            @is_static = is_static
            @return_type = return_type
            @args = args
        end

        def call(obj, args)
            if args.length != @args.length
                raise "Expected #{@args.length} arguments, but got #{args.length} in #{@cls.name}::#{@name} call!";
            end
    
            realObj = @is_static ? @cls.type_obj : obj;
            return realObj.send(@name, *args);
        end
    end

    class MethodArgument
        attr_accessor(:name, :type)

        def initialize(name, type)
            @name = name
            @type = type
        end
    end
end